### PR TITLE
Avoid inhibiting all warning messages for unit tests and fix the warnings.

### PR DESCRIPTION
**Pull request by ashwinstar**
"-w" flag in Makefile.mock inhibits all warning messages. That's not ideal. Hence, remove the flag.
This PR fixes the warnings from unit test which get emitted now.

Unused-functions and unused variables warnings are generated from mock file compilation as `static` functions and variables are never used for mock files. Hence, adding explicit flag to ignore those warnings when compiling mock files only. I tried to skip generating mock versions for static functions and variables, but it needed some fancy regex and sounded confusing not to see `static` things in mock files, hence decided better to just ignore such warnings for mock file compilations.

Commit "Fix warnings for unit tests." is only one which has lot of changes fixing the warnings for unit tests. Rest of the commits are very small to review.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
Just a note, I know this PR fails on CI currently, as it produces different (more) warnings than I locally compiled with due to difference of using `-O0` vs CI using `-O3`. I will fix more of them tomorrow add to the PR. Current commits are still relevant though :-)

File: src/test/unit/cmockery/cmockery.c [hunk 1]
```diff
@@ -204,7 +204,7 @@ static void remove_always_return_values(ListNode * const map_head,
 		const size_t number_of_symbol_names);
 static int check_for_leftover_values(
 		const ListNode * const map_head, const char * const error_message,
-		const size_t number_of_symbol_names);
+		const size_t number_of_symbol_names) __attribute__((format(printf, 2, 0)));
```
Comment (asimrp) [inline src/test/unit/cmockery/cmockery.c]: Do you plan to submit this commit as a patch to cmockery upstream?

Comment (danielgustafsson) [reply src/test/unit/cmockery/cmockery.c]: It's worth trying, but cmockery has been abandoned by Google for years so we shouldn't hold our breath.

File: gpcontrib/pxf/test/pxffilters_test.c [hunk 2]
```diff
@@ -599,13 +608,14 @@ FuncExpr* build_func_expr_operand(List *args, CoercionForm funcformat) {
...
-FuncExpr* build_nested_func_expr_operand(List *columnsOids, List *attrsIndices) {
+static FuncExpr *
+build_nested_func_expr_operand(List *columnsOids, List *attrsIndices) {
 	assert_int_equal(columnsOids->length, attrsIndices->length);
 	ListCell *columnOidLc = NULL, *attrIndexLc = NULL;
 	Var *var = NULL;
 	List* args = NIL;
 	forboth(columnOidLc, columnsOids, attrIndexLc, attrsIndices) {
-		var = build_var(lfirst(columnOidLc), lfirst(attrIndexLc));
+		var = build_var((Oid)lfirst(columnOidLc), (int)lfirst(attrIndexLc));
```
[the remark this answers is not in the document]
Comment (ashwinstar) [reply gpcontrib/pxf/test/pxffilters_test.c]: Thanks a lot, fixed it. In flow of fixing so many warnings didn't remember specific functions existed for the same :)

File: src/backend/replication/test/gp_replication_test.c [hunk 3]
```diff
@@ -6,11 +6,12 @@
...
+#undef Assert
+#define Assert(condition) if (!(condition)) AssertFailed()
 
 bool is_assert_failed = false;
 
-void AssertFailed()
+static void AssertFailed()
 {
 	is_assert_failed = true;
 }
```
[the remark this answers is not in the document]
Comment (ashwinstar) [reply src/backend/replication/test/gp_replication_test.c]: It didn't seem to serve any purpose, deleting the same results in test continue to pass with asserts build. So, deleted the same, thanks for spotting that.

File: src/test/unit/cmockery/cmockery.c (hunk 1 above)
Comment (ashwinstar) [reply src/test/unit/cmockery/cmockery.c]: I tried compiling the latest cmockery and number of warnings generated for it are very high, so don't feel motivated to send patch to fix just few of them (these are not even reported based on gcc flags by default used to compile the project)

**Pull request merged**